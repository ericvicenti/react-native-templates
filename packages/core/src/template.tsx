import React, { ReactNode } from 'react'

export type Store<V = unknown> = {
  get: () => V
  subscribe: (handler: () => void) => () => void
}

export type DataSource = {
  get: (key: string) => Store
}

/** Data */
type DataState = ComponentDataState
// | ReferencedDataState
export enum DataStateType {
  Component = 'component',
  Ref = 'ref',
}

/** Components */
type ComponentIdentifier = string
export type ComponentRegistry = Record<ComponentIdentifier, ComponentDefinition>
export type ComponentDefinition<Props extends object = object> = {
  component: React.ComponentType<
    {
      onTemplateEvent: (name: string, payload: any) => void
      children: ReactNode
    } & Props
  >
  validator?: (input?: ComponentProps) => void
}

// tbd: we should allow objects as props too
type ComponentProp = ComponentProp[] | string | number | boolean | DataState | null | undefined
type ComponentProps = Record<string, ComponentProp>
type ComponentDataState = {
  $: DataStateType.Component
  key: string
  component: ComponentIdentifier
  children?: ComponentProp
  props?: ComponentProps
}

// tbd: needs a better name
export function BaseTemplate({
  components,
  dataState,
  onEvent,
}: {
  components: ComponentRegistry
  dataState: DataState | DataState[]
  onEvent: (key: string, name: string, payload: any) => void
}) {
  function renderComponent(stateNode: ComponentDataState, key: string) {
    const componentDefinition = components[stateNode.component]
    if (!componentDefinition) {
      throw new RenderError(`Unknown component: ${stateNode.component}`)
    }

    // tbd: validate `components` prop once with `zod` instead of later render stage
    const Component = componentDefinition.component
    if (!Component) {
      throw new RenderError(`Invalid component: ${stateNode.component}`)
    }

    if (typeof componentDefinition.validator === 'function') {
      componentDefinition.validator(stateNode.props)
    }

    const props = Object.fromEntries(
      Object.entries(stateNode.props || {}).map(([propKey, propValue]) => {
        return [propKey, render(propValue, `${key}.$props`)]
      })
    )

    const children = stateNode.children ? render(stateNode.children, key) : null

    return (
      <Component
        data-testId={key}
        // tbd: should key be nested with parent, or should it equal to stateNode.key?
        // I guess we use it for communication with the parent, but also react rendering
        key={key}
        children={children}
        {...props}
        onTemplateEvent={(name, payload) => onEvent(key, name, payload)}
      />
    )
  }

  function render(stateNode: ComponentProp, parentKey?: string): React.ReactNode {
    if (stateNode === null || typeof stateNode !== 'object') {
      return stateNode
    }
    if (Array.isArray(stateNode)) {
      return stateNode.map((item) => render(item, parentKey))
    }
    if (stateNode.$ === DataStateType.Component) {
      return (
        <ErrorBoundary>
          {renderComponent(stateNode, parentKey ? `${parentKey}.${stateNode.key}` : stateNode.key)}
        </ErrorBoundary>
      )
    }
    // tbd: what to do with ref
    throw new Error('ref is not supported as a prop yet.')
  }

  return <>{render(dataState)}</>
}

/** Error boundary */
class RenderError extends Error {}
class ErrorBoundary extends React.Component<
  React.PropsWithChildren<object>,
  { error: RenderError | null }
> {
  constructor(props: React.PropsWithChildren<object>) {
    super(props)
    this.state = { error: null }
  }

  static getDerivedStateFromError(error: RenderError) {
    return { error }
  }

  render() {
    if (this.state.error) {
      // tbd: render fallback ui
      return null
    }

    return this.props.children
  }
}
