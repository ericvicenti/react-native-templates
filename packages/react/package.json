{
  "name": "@final-ui/react",
  "version": "0.2.0",
  "license": "Apache-2.0",
  "description": "",
  "main": "lib",
  "react-native": "src",
  "files": [
    "src",
    "lib",
    "!**/__tests__",
    "!**/__fixtures__",
    "!**/__mocks__",
    "!**/.*"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/final-ui/final.git",
    "directory": "packages/react"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "Eric Vicenti <ericvicenti@gmail.com> (https://github.com/ericvicenti)",
  "contributors": [
    "Mike Grabowski <grabbou@gmail.com> (https://github.com/grabbou)"
  ],
  "bugs": {
    "url": "https://github.com/final-ui/final/issues"
  },
  "homepage": "https://github.com/final-ui/final#readme",
  "devDependencies": {
    "@types/react": "^18.0.27",
    "react": "^18.2.0"
  },
  "peerDependencies": {
    "react": "^18.2.0"
  },
  "dependencies": {
    "zod": "^3.22.4"
  }
}
