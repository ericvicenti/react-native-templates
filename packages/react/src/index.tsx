export * from './refs'
export * from './schema'
export * from './template'
