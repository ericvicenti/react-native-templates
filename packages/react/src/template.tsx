import React from 'react'

/** Components */
type ComponentIdentifier = string

export type ComponentRegistry = Record<ComponentIdentifier, ComponentDefinition<any>>
export type ComponentDefinition<T extends Record<string, JSONValue>> = {
  component: React.ComponentType<T>
  validator?: (input?: T) => T
}

/** Data state */
export type DataState = JSONValue | ReferencedComponentDataState
export type ComponentDataState = {
  $: 'component'
  key?: string
  component: ComponentIdentifier
  children?: JSONValue
  props?: Record<string, JSONValue>
}
type ReferencedComponentDataState = ComponentDataState & {
  path: Path
}
export type Path = string | [string, ...(string | number)[]]
export type ReferencedDataState = {
  $: 'ref'
  ref: Path
}
type EventDataState = ActionEventDataState | HandlerEventDataState
export type ActionEventDataState<T = any> = {
  $: 'event'
  action?: T
}
export type HandlerEventDataState = {
  $: 'event'
  key: string
  async: boolean
}
export function isHandlerEvent(
  event: TemplateEvent
): event is TemplateEvent<HandlerEventDataState> {
  return isEventDataState(event.dataState) && 'key' in event.dataState
}
export function isActionEvent(event: TemplateEvent): event is TemplateEvent<ActionEventDataState> {
  return isEventDataState(event.dataState) && 'action' in event.dataState
}

type SafeObject = {
  [key: string]: JSONValue
  $?: never
}
export type JSONValue =
  | ComponentDataState
  | ReferencedDataState
  | EventDataState
  | SafeObject
  | string
  | number
  | boolean
  | null
  | undefined
  | JSONValue[]

export type TemplateEvent<T = EventDataState, K = any> = {
  target: {
    key?: string
    component: string
    propKey: string
    path: Path
  }
  dataState: T
  payload: K
}
export type ActionEvent<T = any, K = any> = TemplateEvent<ActionEventDataState<T>, K>

export function isCompositeDataState(obj: any): obj is ComponentDataState | ReferencedDataState {
  return (
    obj !== null &&
    typeof obj === 'object' &&
    '$' in obj &&
    (obj.$ === 'component' || obj.$ === 'ref')
  )
}
export function isComponentDataState(obj: JSONValue): obj is ComponentDataState {
  return obj !== null && typeof obj === 'object' && '$' in obj && obj.$ === 'component'
}
export function isReferencedComponentDataState(
  obj: JSONValue
): obj is ReferencedComponentDataState {
  return isComponentDataState(obj) && 'path' in obj
}
export function isEventDataState(obj: JSONValue): obj is EventDataState {
  return obj !== null && typeof obj === 'object' && '$' in obj && obj.$ === 'event'
}

export function BaseTemplate({
  path = '',
  components,
  dataState,
  onEvent,
}: {
  path?: Path
  components: ComponentRegistry
  dataState: JSONValue
  onEvent?: (event: TemplateEvent) => Promise<any>
}) {
  function renderComponent(stateNode: ComponentDataState, path: Path) {
    const componentDefinition = components[stateNode.component]
    if (!componentDefinition) {
      throw new RenderError(`Unknown component: ${stateNode.component}`)
    }

    // tbd: validate `components` prop once with `zod` instead of later render stage
    const Component = componentDefinition.component
    if (!Component) {
      throw new RenderError(`Invalid component: ${stateNode.component}`)
    }

    let componentProps = Object.fromEntries(
      Object.entries(stateNode.props || {}).map(([propKey, propValue]) => {
        return [propKey, renderProp(propKey, propValue, stateNode, path)]
      })
    )

    if (typeof componentDefinition.validator === 'function') {
      try {
        componentProps = componentDefinition.validator(componentProps)
      } catch (e) {
        throw new RenderError(
          `Invalid props for component: ${stateNode.component}, props: ${JSON.stringify(stateNode.props)}. Error: ${JSON.stringify(e)}`
        )
      }
    }

    const children = stateNode.children ? render(stateNode.children, path) : null

    return <Component key={stateNode.key} {...componentProps} children={children} />
  }

  function render(stateNode: JSONValue, path: Path): React.ReactNode {
    if (stateNode === null || typeof stateNode !== 'object') {
      return stateNode
    }
    if (Array.isArray(stateNode)) {
      return stateNode.map((item) => render(item, path))
    }
    if (!isCompositeDataState(stateNode)) {
      throw new Error('Objects are not valid as a React child.')
    }
    if (stateNode.$ === 'component') {
      return renderComponent(
        stateNode,
        isReferencedComponentDataState(stateNode) ? stateNode.path : path
      )
    }
    if (stateNode.$ === 'ref') {
      throw new Error('Your data includes refs. You must use a <Template /> component instead.')
    }
  }

  function renderProp(
    propKey: string,
    propValue: JSONValue,
    parentNode: ComponentDataState | ReferencedComponentDataState,
    path: Path
  ): any {
    if (isEventDataState(propValue)) {
      return async (payload: any) => {
        // React events (e.g. from onPress) contain cyclic structures that can't be serialized
        // with JSON.stringify and also provide little to no value for the server.
        // tbd: figure a better way to handle this in a cross-platform way
        if (payload?.nativeEvent) {
          payload = '[native code]'
        }
        return onEvent?.({
          target: {
            key: parentNode.key,
            component: parentNode.component,
            propKey,
            path,
          },
          dataState: propValue,
          payload,
        })
      }
    }
    if (Array.isArray(propValue)) {
      return propValue.map((item) => renderProp(propKey, item, parentNode, path))
    }
    if (isCompositeDataState(propValue)) {
      return render(propValue, path)
    }
    if (propValue && typeof propValue === 'object') {
      return Object.fromEntries(
        Object.entries(propValue).map(([key, value]) => {
          return [key, renderProp(key, value, parentNode, path)]
        })
      )
    }
    return propValue
  }

  return <ErrorBoundary>{render(dataState, path)}</ErrorBoundary>
}

/** Error boundary */
class RenderError extends Error {}
class ErrorBoundary extends React.Component<
  React.PropsWithChildren<object>,
  { error: RenderError | null }
> {
  constructor(props: React.PropsWithChildren<object>) {
    super(props)
    this.state = { error: null }
  }

  static getDerivedStateFromError(error: RenderError) {
    return { error }
  }

  render() {
    if (this.state.error) {
      // tbd: render fallback ui
      return null
    }

    return this.props.children
  }
}
