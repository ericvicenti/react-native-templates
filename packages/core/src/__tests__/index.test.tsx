/**
 * @jest-environment jsdom
 */

import { render } from '@testing-library/react'
import React from 'react'

import { BaseTemplate, ComponentDefinition, ComponentRegistry, DataStateType } from '../template'

const BUILT_IN_COMPONENTS: ComponentRegistry = {
  View: {
    component: (props) => <div {...props} />,
  },
}

it('should render a component', () => {
  const component = render(
    <BaseTemplate
      components={BUILT_IN_COMPONENTS}
      dataState={{
        $: DataStateType.Component,
        key: '$root',
        component: 'View',
        props: {
          height: 50,
        },
      }}
      onEvent={jest.fn()}
    />
  )

  expect(component.getByTestId('$root')).toMatchSnapshot()
})

it('should render an array of components', () => {
  const component = render(
    <BaseTemplate
      components={BUILT_IN_COMPONENTS}
      dataState={[
        {
          $: DataStateType.Component,
          key: '$heading',
          component: 'View',
          children: 'Heading',
        },
        {
          $: DataStateType.Component,
          key: '$paragraph',
          component: 'View',
          children: 'Hello, world!',
        },
      ]}
      onEvent={jest.fn()}
    />
  )

  expect(component.asFragment()).toMatchSnapshot()
})

it('should render a component with children of different types', () => {
  const component = render(
    <BaseTemplate
      components={BUILT_IN_COMPONENTS}
      dataState={[
        {
          $: DataStateType.Component,
          key: '$heading',
          component: 'View',
          children: [
            'Hello',
            {
              $: DataStateType.Component,
              key: '$nested',
              component: 'View',
              children: 'world!',
            },
          ],
        },
      ]}
      onEvent={jest.fn()}
    />
  )

  expect(component.asFragment()).toMatchSnapshot()
})

it('should accept component as a prop', () => {
  const Header: ComponentDefinition<{ header: string; paragraph: string }> = {
    component: (props) => (
      <section>
        <header>{props.header}</header>
        <footer>{props.paragraph}</footer>
      </section>
    ),
  }

  const component = render(
    <BaseTemplate
      components={{
        ...BUILT_IN_COMPONENTS,
        // @ts-ignore figure why this type does not work
        Header,
      }}
      dataState={[
        {
          $: DataStateType.Component,
          key: '$heading',
          component: 'Header',
          props: {
            header: {
              $: DataStateType.Component,
              key: '$header',
              component: 'View',
              children: 'Header text',
            },
            paragraph: {
              $: DataStateType.Component,
              key: '$footer',
              component: 'View',
              children: 'Footer text',
            },
          },
        },
      ]}
      onEvent={jest.fn()}
    />
  )

  expect(component.asFragment()).toMatchSnapshot()
})
