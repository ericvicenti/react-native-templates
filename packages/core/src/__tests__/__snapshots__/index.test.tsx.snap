// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept component as a prop 1`] = `
<DocumentFragment>
  <section>
    <header>
      <div
        data-testid="$heading.$props.$header"
      >
        Header text
      </div>
    </header>
    <footer>
      <div
        data-testid="$heading.$props.$footer"
      >
        Footer text
      </div>
    </footer>
  </section>
</DocumentFragment>
`;

exports[`should render a component 1`] = `
<div
  data-testid="$root"
  height="50"
/>
`;

exports[`should render a component with children of different types 1`] = `
<DocumentFragment>
  <div
    data-testid="$heading"
  >
    Hello
    <div
      data-testid="$heading.$nested"
    >
      world!
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render an array of components 1`] = `
<DocumentFragment>
  <div
    data-testid="$heading"
  >
    Heading
  </div>
  <div
    data-testid="$paragraph"
  >
    Hello, world!
  </div>
</DocumentFragment>
`;
